src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
-        set_result(txAndResult.second);
-        return true;
+        builder_.AddSaplingSpend(saplingKeys[i].expsk, saplingNotes[i], anchor, witnesses[i].value());
     }
-    /**
-     * END SCENARIO #1
-     */
-
-
-    // Prepare raw transaction to handle JoinSplits
-    CMutableTransaction mtx(tx_);
-    ed25519_generate_keypair(&joinSplitPrivKey_, &joinSplitPubKey_);
-    mtx.joinSplitPubKey = joinSplitPubKey_;
-    tx_ = CTransaction(mtx);
-
-    // Copy zinputs and zoutputs to more flexible containers
-    std::deque<SendManyInputJSOP> zInputsDeque; // zInputsDeque stores minimum numbers of notes for target amount
-    CAmount tmp = 0;
-    for (auto o : z_sprout_inputs_) {
-        zInputsDeque.push_back(o);
-        tmp += o.amount;
-        if (tmp >= txValues.targetAmount) {
+
+    // Add Sapling and transparent outputs
+    for (const auto& r : recipients_) {
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                builder_.AddTransparentOutput(keyId, r.amount);
+            },
+            [&](const CScriptID& scriptId) {
+                builder_.AddTransparentOutput(scriptId, r.amount);
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                //unreachable
+                throw JSONRPCError(
+                    RPC_INVALID_ADDRESS_OR_KEY,
+                    "Sending funds to Sprout is disabled.");
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                auto value = r.amount;
+                auto memo = get_memo_from_hex_string(r.memo.has_value() ? r.memo.value() : "");
+
+                builder_.AddSaplingOutput(ovk, addr, value, memo);
+            },
+            [&](const libzcash::UnifiedAddress& addr) {
+                //unreachable
+                throw JSONRPCError(
+                    RPC_INVALID_ADDRESS_OR_KEY,
+                    "Unified addresses are not yet supported by z_sendmany");
+            }
+        }, r.address);
+    }
+
+    // Add transparent utxos
+    for (const auto& out : spendable.utxos) {
+        const CTxOut& txOut = out.tx->vout[out.i];
+        builder_.AddTransparentInput(COutPoint(out.tx->GetHash(), out.i), txOut.scriptPubKey, txOut.nValue);
+
+        sum += txOut.nValue;
+        if (sum >= txValues.targetAmount) {
             break;
         }
     }
-    std::deque<SendManyRecipient> zOutputsDeque;
-    for (auto o : z_outputs_) {
-        zOutputsDeque.push_back(o);
-    }
 
+    // Find Sprout witnesses
     // When spending notes, take a snapshot of note witnesses and anchors as the treestate will
     // change upon arrival of new blocks which contain joinsplit transactions.  This is likely
     // to happen as creating a chained joinsplit transaction can take longer than the block interval.
-    if (z_sprout_inputs_.size() > 0) {
+    // So, we need to take locks on cs_main and pwalletMain->cs_wallet so that the witnesses aren't
+    // updated.
```

Comment:
The fact that we lock and unlock `cs_main` and `cs_wallet` multiple times, means that the Sprout and Sapling anchors can come from different blocks. This is technically a wallet implementation distinguisher. I don't think it's a big deal in practice, but we could do these in a single `lock/unlock`.

Reply:
We can fix this in a subsequent refactor, after this PR (and probably after extracting common coin selection logic from the three main RPCs).

Reply:
Please add a TODO for that.